crates/sui-types/src/signature_verification.rs

```diff
@@ -5,15 +5,111 @@ use nonempty::NonEmpty;
 use crate::transaction::{SenderSignedData, TransactionDataAPI};
+use lru::LruCache;
+use parking_lot::RwLock;
+use prometheus::IntCounter;
+use std::hash::Hash;
+use std::sync::Arc;
+
+// Cache up to 20000 verified certs. We will need to tune this number in the future - a decent
+// guess to start with is that it should be 10-20 times larger than peak transactions per second,
+// on the assumption that we should see most certs twice within about 10-20 seconds at most: Once via RPC, once via consensus.
+const VERIFIED_CERTIFICATE_CACHE_SIZE: usize = 20000;
+
+pub struct VerifiedDigestCache<D> {
+    inner: RwLock<LruCache<D, ()>>,
+    cache_hits_counter: IntCounter,
+    cache_misses_counter: IntCounter,
+    cache_evictions_counter: IntCounter,
+}
+
+impl<D: Hash + Eq + Copy> VerifiedDigestCache<D> {
+    pub fn new(
+        cache_hits_counter: IntCounter,
+        cache_misses_counter: IntCounter,
+        cache_evictions_counter: IntCounter,
+    ) -> Self {
+        Self {
+            inner: RwLock::new(LruCache::new(
+                std::num::NonZeroUsize::new(VERIFIED_CERTIFICATE_CACHE_SIZE).unwrap(),
+            )),
+            cache_hits_counter,
+            cache_misses_counter,
+            cache_evictions_counter,
+        }
+    }
+
+    pub fn is_cached(&self, digest: &D) -> bool {
+        let inner = self.inner.read();
+        if inner.contains(digest) {
+            self.cache_hits_counter.inc();
+            true
+        } else {
+            self.cache_misses_counter.inc();
+            false
+        }
+    }
+
+    pub fn cache_digest(&self, digest: D) {
+        let mut inner = self.inner.write();
+        if let Some(old) = inner.push(digest, ()) {
+            if old.0 != digest {
+                self.cache_evictions_counter.inc();
+            }
+        }
+    }
+
+    pub fn cache_digests(&self, digests: Vec<D>) {
+        let mut inner = self.inner.write();
+        digests.into_iter().for_each(|d| {
+            if let Some(old) = inner.push(d, ()) {
+                if old.0 != d {
+                    self.cache_evictions_counter.inc();
+                }
+            }
+        });
+    }
+
+    pub fn is_verified<F, G>(&self, digest: D, verify_callback: F, uncached_checks: G) -> SuiResult
+    where
+        F: FnOnce() -> SuiResult,
+        G: FnOnce() -> SuiResult,
+    {
+        if !self.is_cached(&digest) {
+            verify_callback()?;
+            self.cache_digest(digest);
+        } else {
+            // Checks that are required to be performed outside the cache.
+            uncached_checks()?;
+        }
+        Ok(())
+    }
+
+    pub fn clear(&self) {
+        let mut inner = self.inner.write();
+        inner.clear();
+    }
+
+    // Initialize an empty cache when the cache is not needed (in testing scenarios, graphql and rosetta initialization).
+    pub fn new_empty() -> Self {
+        Self::new(
+            IntCounter::new("test_cache_hits", "test cache hits").unwrap(),
```

Comment:
It's not clear to me if this type of counter initialization would actually work in prod (we only use this style in tests). Specifically, if you happen to have two of this instances that have the same name, would it cause runtime issues?

Reply:
i did not find any uniqueness constraints on the counter names anywhere - most of the new_empty is invoked from tests only, i guess the question is for graphql and whether they have multiple instances?

Reply:
one of the options is to add a random prefix to the counter name